### ISSUE TITLE
Ensure style guide documentation is complete

### ISSUE THREAD

**Issue by rt17603** [openghg/openghg#782]
There is a useful [Coding Style](https://github.com/openghg/openghg/blob/devel/doc/source/development/python_devel.rst#coding-style) included in the documentation. To ensure we keep a consistent style this should be updated to include anything which would be fed back on a pull request. This includes:

 - Doc-string formatting
 - Argument specification for internal calls to functions

Before this is finalised this should be presented to the OpenGHG team for final approval.

**Comment by brendan-m-murphy** [comment]
Some points around typing:

- with `from future import __annotations__` we can use `dict, tuple, list` directly instead of importing them from `typing`
- if the type in a `list` or `dict` doesn't matter, we should use generics:
  ```
  T = TypeVar("T")
  def fn(stuff: list[T]) -> list[T]:
      ...
  def get_values(keyed_stuff: dict[str, T]) -> list[T]:
      return list(keyed_stuff.values())
  ```
  For some functions in `store`, they take `Dict` as an argument, but it's actually more like `dict[str, dict[str, Any]]` (and maybe should be `dict[str, tuple[MetaData, xr.Dataset, Optional[MetaData]]]`, where `MetaData = dict[str, Any]`).

Related to the last point is the use of type aliases like `MetaData = dict[str, Any]`. This is bad for the API (standardise, etc.) since the docstring will be less useful to a user. But for internal code, it signifies intent more than `dict[str, Any]` does (I think), and opens the option to make a `MetaData` class (for instance). So, I think they have some merit for cases where the same type hint and variable name almost always go together and they are used in a specific way (but maybe it's not worth making a proper class for them yet).

Some points from the [google style guide](https://google.github.io/styleguide/pyguide.html#doc-function-args) we might want to consider:
- is `Returns: None` necessary?
- If the first line of the docstring specifies the return value well enough, do we need "Returns"?
- Do we need to specify the type with "Returns" if there is a type hint?

Also, our current style is: new line before "Args", no new line before "Returns"

Maybe for the PR template, I will make a checkbox for docstring formatting, since I seem to forget it a lot.

**Comment by gareth-j** [comment]
Yeah I think we should always have a Returns, it makes glancing at a function docstring easy. Specifiying the return type in the docstring is easy and having Returns means we can add a bit more info about the object that is returned. I know the user can probably work out what it's returning from the function name and return type but sometimes it's good to have a bit of explanation.

**Comment by rt17603** [comment]
> with from future import __annotations__ we can use dict, tuple, list directly instead of importing them from typing

Probably worth avoiding this if possible based on - https://docs.python.org/3/library/__future__.html#id1. I don't think it's a problem to use the typing module for now.

**Comment by rt17603** [comment]
> if the type in a list or dict doesn't matter, we should use generics
Yes I agree with this.

**Comment by rt17603** [comment]
> Maybe for the PR template, I will make a checkbox for docstring formatting, since I seem to forget it a lot.
Yes I think that sounds like a good idea - and we can add a link there to this updated style guide!

**Comment by brendan-m-murphy** [comment]
> > with `from future import __annotations__` we can use dict, tuple, list directly instead of importing them from typing
> 
> Probably worth avoiding this if possible based on - https://docs.python.org/3/library/__future__.html#id1. I don't think it's a problem to use the `typing` module for now.

I think it is unavoidable in cases where we want to use `TypeVar` before the class is defined (e.g. with class methods): https://peps.python.org/pep-0563/#forward-references

Also using `dict`, `list`, `tuple` is in Python 3.9, just not in 3.8. (But I'm just being lazy and don't want to import `List` etc...)

**Comment by rt17603** [comment]
From the linked article:

> from __future__ import annotations was previously scheduled to become mandatory in Python 3.10, but the Python Steering Council twice decided to delay the change ([announcement for Python 3.10](https://mail.python.org/archives/list/python-dev@python.org/message/CLVXXPQ2T2LQ5MP2Y53VVQFCXYWQJHKZ/); [announcement for Python 3.11](https://mail.python.org/archives/list/python-dev@python.org/message/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/)). No final decision has been made yet.

I don't think we should be using something which is divisive like this if we can avoid it (which it seems like we can).

**Comment by brendan-m-murphy** [comment]
Another thing to consider is class and module docstrings. I think module level docstrings can be useful to describe the scope of what the module is trying to accomplish. I think maybe google suggests listing all public methods? But that doesn't necessarily work with our module set-up.

For the module set-up, we have a lot of circular imports that Python ignores, e.g. if module A and module B are in the same subpackage P, if module A contains `from P import something_from_B` then that's technically circular. You can avoid it by using `from .B import something_from_B`.

It might also help with the overall structure if we put all imports at the top level of modules. At the moment, it is hiding some circular imports. Once or twice I've tried to add code somewhere that seemed logical (`store.spec` I think), and had circular imports (from `objectstore._integrity`) that meant the imports had to be moved inside a function.

The first circular import thing doesn't really matter, but I think the second one might suggest that the code needs to be reorganised (which is partially why the new metastore stuff is in `objectstore` and not in `store`). I'm not saying all imports inside functions are bad, but I think it's better to avoid them unless they're used to intentionally delay loading a large package, or something like that.

**Comment by rt17603** [comment]
For this, could we start by just getting in some details about doc-string syntax first? Seems like this is the current question that's come up a few times and would be good to get sorted.

These other questions seem more broad and need a bigger discussion.

**Comment by brendan-m-murphy** [comment]
> From the linked article:
> 
> > from `__future__ import annotations` was previously scheduled to become mandatory in Python 3.10, but the Python Steering Council twice decided to delay the change ([announcement for Python 3.10](https://mail.python.org/archives/list/python-dev@python.org/message/CLVXXPQ2T2LQ5MP2Y53VVQFCXYWQJHKZ/); [announcement for Python 3.11](https://mail.python.org/archives/list/python-dev@python.org/message/VIZEBX5EYMSYIJNDBF6DMUMZOCWHARSO/)). No final decision has been made yet.
> 
> I don't think we should be using something which is divisive like this if we can avoid it (which it seems like we can).

We can avoid it for `list`, `dict`, etc. but if we don't use it for declarations like `T = TypeVar('T', bound='Datasource')`, then we can't have class methods and require type annotations on all functions.

It does seem like we'll need to be careful if want to use pydantic. It has a lot of features that would make it useful for managing metadata (like validation over multiple fields, like 'inlet' and 'height').

So is it fair to say:

If `from future import __annotations__` is necessary for a forward type hint (e.g. for class methods), then I guess there is no harm in using `list` etc. for typing, since that is supported in 3.9.

If we don't need `future.__annotations__` for class methods (or other special situations), then we should import `List` etc from `typing` as long as we support 3.8.

Or should we always use `List` over `list` (and so on) while 3.8 is supported?

**Comment by brendan-m-murphy** [comment]
> For this, could we start by just getting in some details about doc-string syntax first? Seems like this is the current question that's come up a few times and would be good to get sorted.
> 
> These other questions seem more broad and need a bigger discussion.

Okay, I think the only docstring things that haven't been addressed are:
- blank lines before "Args", "Returns" and "Raises", or just before "Args"
- module and class level docstrings (e.g. to define the scope/responsibility/purpose of the module/class)

**Comment by brendan-m-murphy** [comment]
And so far we have:
- always include "Returns" with the return type in docstrings
- add docstring formatting to PR template with link to docs

**Comment by brendan-m-murphy** [comment]
Another thing I'm not sure about is when the description for an argument or for "Returns" runs over a line. I've been indenting one indent past "Args:" or "Returns:" recently.

**Comment by brendan-m-murphy** [comment]
Note: pylint will remind you of missing  docstrings
